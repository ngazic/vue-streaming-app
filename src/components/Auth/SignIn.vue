<template>
<div id="signin">
  <div class="signin-form">
    <form @submit.prevent="onSubmit">
      <div class="input">
        <label for="user">User</label>
        <input type="user" id="user" v-model="user" />
      </div>
      <div class="input">
        <label for="password">Password</label>
        <input type="password" id="password" v-model="password" />
      </div>
      <div class="submit">
        <button type="submit">Submit</button>
      </div>
    </form>
  </div>
</div>
</template>

<script lang="ts">
import Vue from "vue";
import {
  mapActions
} from "vuex";
export default Vue.extend({
  data() {
    return {
      user: "",
      password: ""
    };
  },
  methods: {
    ...mapActions(["login"]),
    onSubmit() {
      const data: {
        email: string;
        password: string;
      } = {
        email: this.user,
        password: this.password
      };
      this.login(data)
        .then(() => {
          if (this.$store.getters.isAuthenticated) {
            // eslint-disable-next-line no-console
            console.log("change route");
            this.$router.push("/");
          }
        })
        .catch(err => alert(err));
    }
  }
});
</script>

<style scoped>
.signin-form {
  max-width: 400px;
  margin: 30px auto;
  border: 1px solid #eee;
  padding: 20px;
  box-shadow: 0 2px 3px #ccc;
}

.input {
  margin: 10px auto;
}

.input label {
  display: block;
  color: #4e4e4e;
  margin-bottom: 6px;
}

.input input {
  font: inherit;
  width: 100%;
  padding: 6px 12px;
  box-sizing: border-box;
  border: 1px solid #ccc;
}

.input input:focus {
  outline: none;
  border: 1px solid #521751;
  background-color: #eee;
}

.submit button {
  border: 1px solid #521751;
  color: #521751;
  padding: 10px 20px;
  font: inherit;
  cursor: pointer;
}

.submit button:hover,
.submit button:active {
  background-color: #521751;
  color: white;
}

.submit button[disabled],
.submit button[disabled]:hover,
.submit button[disabled]:active {
  border: 1px solid #ccc;
  background-color: transparent;
  color: #ccc;
  cursor: not-allowed;
}
</style>
