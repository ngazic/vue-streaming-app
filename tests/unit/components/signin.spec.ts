import { shallowMount, createLocalVue, Wrapper } from '@vue/test-utils'
import Vuex, { StoreOptions } from 'vuex'
import VueRouter from 'vue-router'
import SignIn from '@/components/Auth/SignIn.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)
const router = new VueRouter()
jest.spyOn(window, 'alert').mockImplementation(() => { return });

describe('SignIn.vue', () => {

  const MockFormData = {
    user: 'John Doe',
    password: 'secret1234'
  }

  let wrapper: Wrapper<Vue>
  let user: Wrapper<Vue>
  let password: Wrapper<Vue>
  let form: Wrapper<Vue>

  function createConfig(overrides: StoreOptions<unknown>): Wrapper<Vue> {
    const store = new Vuex.Store(overrides)

    const wrapper = shallowMount(SignIn, {
      localVue,
      store,
      router
    })
    user = wrapper.find('#user')
    password = wrapper.find('#password')
    form = wrapper.find('form')
    user.setValue(MockFormData.user)
    password.setValue(MockFormData.password)
    return wrapper
  }

  it('renders correctly', () => {
    wrapper = shallowMount(SignIn)
    expect(wrapper).toMatchSnapshot()
  });

  it('dispatches "login" action when form is submited', async () => {
    const actions = {
      login: jest.fn()
    }



    wrapper = createConfig({ actions });
    await form.trigger('submit.prevent')
    expect(actions.login).toHaveBeenCalled()

  });

  it('router called on successful login ', async () => {
    const config = {
      actions: {
        login: jest.fn().mockResolvedValue({})
      }
      , getters: {
        isAuthenticated: (): boolean  => true
      }
    }

    wrapper = createConfig(config);
    const spy = jest.spyOn(wrapper.vm.$router, 'push')
    await form.trigger('submit.prevent')
    expect(spy).toBeCalled()
  });


  it('should show allert message on the REST API error', async () => {
    const actions = {
      login: jest.fn().mockRejectedValue({ err: 'error message' })
    }
    wrapper = createConfig({ actions })
    const spy = jest.spyOn(window, 'alert')
    await form.trigger('submit.prevent')
    expect(spy).toHaveBeenCalledWith({ err: 'error message' })
  });


});