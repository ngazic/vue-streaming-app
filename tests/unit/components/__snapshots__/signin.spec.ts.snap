// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn.vue renders correctly 1`] = `
<div id="signin">
  <div class="signin-form">
    <form>
      <div class="input"><label for="user">User</label> <input type="user" id="user"></div>
      <div class="input"><label for="password">Password</label> <input type="password" id="password"></div>
      <div class="submit"><button type="submit">Submit</button></div>
    </form>
  </div>
</div>
`;
